optuna/optuna
# Integrate multi-objective optimization module

## Motivation
In the current Optuna's codebase, the `optuna.multi_objective` module is separated from other modules. However, to improve maintenance and ensure code consistency, Optuna's basic modules (e.g. study, trial, etc) should be designed with multi-objective optimization in mind. This PR aims to integrate the multi-objective optimization module.

## Description of the changes (Overview)

About `optuna.study`
- Allow `Study.direction` can be multi-dimensional.
- Add `Study.get_pareto_front_trials()`.

About `optuna.storages`
- Add `StudyDirectionModel` because direction can be multi-dimensional.
- Make `TrialValueModel` which corresponds to `Trial.value`.
- Make `TrialInternediateValueModel` which corresponds to `Trial.intermediate_values`.
- Fix each storage to handle multi-dimensional `Study.direction` and `Trial.{value, intermediate_values}`.

About `optuna.samplers`
- Add `optuna.samplers.multi_objective.NSGAIISampler`.

About `optuna.trial`
- Allow the `value` argument of `Trial.report` can be multi-dimensional.
- Unify treatment of `step` argument in `Trial.report` as single-objective and multi-objective.
- Allow `FrozenTral.value` and the element of `FrozenTrial.intermediate_values` could be multi-dimensional.

About `optuna.visualization`
- Add `optuna.visualization.plot_pareto_front()`.

About `optuna.multi-objective`
- Leave all existing code, but deprecate `optuna.multi_objective.{study, trial, samplers, visualization}`.

About `tests`
- Create a test with the current multi-objective optimization module test case as it is.
- Add tests in storages for multiple directions, value, and intermediate_values.

Other
- Add several assertions to pass `mypy`'s check in various places.
